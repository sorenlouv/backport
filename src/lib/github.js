const axios = require('axios');
const querystring = require('querystring');
const get = require('lodash.get');
const constants = require('./constants');

let accessToken;
function getCommitMessage(message) {
  return message.split('\n')[0].trim();
}

function getCommits(owner, repoName, author) {
  const urlArgs = {
    per_page: 20,
    access_token: accessToken
  };

  if (author) {
    urlArgs.author = author;
    urlArgs.per_page = 5;
  }

  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
      urlArgs
    )}`
  )
    .catch(handleError)
    .then(res =>
      res.data.map(commit => {
        const message = getCommitMessage(commit.commit.message);
        return {
          message,
          sha: commit.sha
        };
      })
    );
}

function getCommit(owner, repoName, sha) {
  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => ({
      message: res.data.commit.message,
      sha: res.data.sha
    }));
}

function createPullRequest(owner, repoName, payload) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    )
    .catch(handleError);
}

function addLabels(owner, repoName, pullNumber, labels) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    )
    .catch(handleError);
}

function getPullRequestByCommit(owner, repoName, commitSha) {
  return axios(
    `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}&access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => get(res.data.items[0], 'number'));
}

function setAccessToken(_accessToken) {
  accessToken = _accessToken;
}

class GithubError extends Error {
  constructor(message) {
    super();
    Error.captureStackTrace(this, GithubError);
    this.code = constants.GITHUB_ERROR;
    this.message = message;
  }
}

function handleError(e) {
  if (get(e.response, 'data')) {
    throw new GithubError(e.response.data);
  }

  throw e;
}

module.exports = {
  setAccessToken,
  addLabels,
  createPullRequest,
  getCommit,
  getCommits,
  getPullRequestByCommit
};
