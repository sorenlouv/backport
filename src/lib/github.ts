import {
  GithubQuery,
  GithubCommit,
  PullRequest,
  Commit,
  GithubIssue,
  GithubPullRequestPayload,
  GithubIssues,
  GithubApiError
} from '../types/types';

import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import { HandledError } from './errors';

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function getCommits(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 20
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullRequest: await getPullRequestBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function getCommit(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubCommit> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
    );
    const fullSha = res.data.sha;
    const pullRequest = await getPullRequestBySha(owner, repoName, fullSha);

    return {
      message: getCommitMessage(res.data.commit.message),
      sha: fullSha,
      pullRequest
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: GithubPullRequestPayload
): Promise<PullRequest> {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabels(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

async function getPullRequestBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubIssues> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
