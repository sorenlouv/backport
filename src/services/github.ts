import {
  GithubApiError,
  GithubCommit,
  GithubIssue,
  GithubQuery,
  GithubSearch
} from '../types/GithubApi';
import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import isEmpty from 'lodash.isempty';
import { HandledError } from './HandledError';
import { getPullRequestPayload } from '../steps/doBackportVersions';

export interface Commit {
  sha: string;
  message: string;
  pullNumber?: number;
}

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function fetchCommitsByAuthor(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 10
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullNumber: await fetchPullRequestNumberBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function fetchCommitBySha(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubSearch<GithubCommit>> = await axios(
      `https://api.github.com/search/commits?q=hash:${sha}%20repo:${owner}/${repoName}&per_page=1&access_token=${accessToken}`,
      {
        headers: {
          Accept: 'application/vnd.github.cloak-preview'
        }
      }
    );

    if (isEmpty(res.data.items)) {
      throw new HandledError(`No commit found for SHA: ${sha}`);
    }

    const commitRes = res.data.items[0];
    const fullSha = commitRes.sha;
    const pullNumber = await fetchPullRequestNumberBySha(
      owner,
      repoName,
      fullSha
    );

    return {
      message: getCommitMessage(commitRes.commit.message),
      sha: fullSha,
      pullNumber
    };
  } catch (e) {
    throw getError(e);
  }
}

async function fetchPullRequestNumberBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubSearch<GithubIssue>> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: ReturnType<typeof getPullRequestPayload>
) {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabelsToPullRequest(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
