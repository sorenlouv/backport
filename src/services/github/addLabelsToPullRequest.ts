import axios from 'axios';
import { BackportOptions } from '../../options/options';
import { handleGithubError } from './handleGithubError';
import { logger } from '../logger';

export async function addLabelsToPullRequest(
  { apiHostname, repoName, repoOwner, accessToken, username }: BackportOptions,
  pullNumber: number,
  labels: string[]
) {
  logger.info(`Adding label "${labels}" to #${pullNumber}`);

  try {
    return await axios.post(
      `https://${apiHostname}/repos/${repoOwner}/${repoName}/issues/${pullNumber}/labels`,
      labels,
      {
        auth: {
          username: username,
          password: accessToken
        }
      }
    );
  } catch (e) {
    throw handleGithubError(e);
  }
}
