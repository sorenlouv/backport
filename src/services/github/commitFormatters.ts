export function getShortSha(sha: string) {
  return sha.slice(0, 8);
}

export function getFirstCommitMessageLine(message: string) {
  return message.split('\n')[0];
}

export function getFormattedCommitMessage({
  message,
  pullNumber,
  sha,
}: {
  message: string;
  pullNumber?: number;
  sha: string;
}) {
  const firstMessageLine = getFirstCommitMessageLine(message);
  const messageHasPullNumber = firstMessageLine.match(/.+ \(#\d+\)/);

  // message already contains pull number
  if (messageHasPullNumber) {
    return firstMessageLine;
  }

  // message doesn't contain pull number. Add it
  if (pullNumber) {
    return `${firstMessageLine} (#${pullNumber})`;
  }

  // pull number not available. Add commit
  return `${firstMessageLine} (${getShortSha(sha)})`;
}

export function getPullNumberFromMessage(firstMessageLine: string) {
  const matches = firstMessageLine.match(/\(#(\d+)\)/);
  if (matches) {
    return parseInt(matches[1], 10);
  }
}
