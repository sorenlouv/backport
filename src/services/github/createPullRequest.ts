import axios, { AxiosResponse } from 'axios';
import { BackportOptions } from '../../options/options';
import { GithubIssue } from './GithubApiTypes';
import { handleGithubError } from './handleGithubError';
import { logger } from '../logger';

export async function createPullRequest(
  { apiHostname, repoName, repoOwner, accessToken }: BackportOptions,
  payload: {
    title: string;
    body: string;
    head: string;
    base: string;
  }
) {
  logger.info(
    `Creating PR with title: "${payload.title}". ${payload.head} -> ${payload.base}`
  );

  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://${apiHostname}/repos/${repoOwner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw handleGithubError(e);
  }
}
