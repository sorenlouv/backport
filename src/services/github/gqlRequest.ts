import axios from 'axios';
import get from 'lodash.get';
import { HandledError } from '../HandledError';
import { logger } from '../logger';

interface GithubResponse<DataResponse> {
  data: DataResponse;
  errors: {
    type: string;
    path: string[];
    locations: {
      line: number;
      column: number;
    }[];
    message: string;
  }[];
}

export async function gqlRequest<DataResponse>({
  apiHostname,
  query,
  variables,
  accessToken
}: {
  apiHostname: string;
  query: string;
  variables?: {
    [key: string]: string | number | null;
  };
  accessToken: string;
}) {
  try {
    logger.verbose(query);
    logger.verbose(variables as any);
    const { data } = await axios.post<GithubResponse<DataResponse>>(
      `https://${apiHostname}/graphql`,
      { query, variables },
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${accessToken}`
        }
      }
    );

    logger.verbose(data);

    if (data.errors) {
      const message = data.errors.map(error => error.message).join(', ');
      throw new HandledError(message);
    }

    return data.data;
  } catch (e) {
    logger.info(e.message);
    logger.info(e.response);
    const responseError = get(e, 'response.data.errors') as {
      message: string;
    }[];
    if (responseError) {
      throw new HandledError(
        responseError.map(error => error.message).join(', ')
      );
    }

    throw e;
  }
}
