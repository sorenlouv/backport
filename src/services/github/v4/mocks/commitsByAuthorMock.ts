import { DataResponse } from '../fetchCommitsByAuthor';

export const commitsWithPullRequestsMock: DataResponse = {
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '2e63475c483f7844b0f2833bc57fdee32095bacb',
                message: 'Add ðŸ‘»',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                message: 'Add witch (#85)',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-b' }],
                        },
                        mergeCommit: {
                          oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                          message: 'Add witch (#85)',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 85,
                        timelineItems: {
                          edges: [],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        baseRefName: 'master',
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '3827bbbaf39914eda4f02f6940189844375fd097',
                message: 'Add backport config',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: '5ea0da550ac191029459289d67f99ad7d310812b',
                message: 'Initial commit',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
          ],
        },
      },
    },
  },
};
