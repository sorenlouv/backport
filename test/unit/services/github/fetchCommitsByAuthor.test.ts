import * as gqlRequest from '../../../../src/services/github/gqlRequest';
import { CommitSelected } from '../../../../src/services/github/Commit';
import {
  PullRequestEdge,
  fetchCommitsByAuthor,
  getExistingBackportPRs
} from '../../../../src/services/github/fetchCommitsByAuthor';
import { commitsWithPullRequestsMock } from './mocks/commitsByAuthorMock';
import { getDefaultOptions } from './getDefaultOptions';

describe('fetchCommitsByAuthor', () => {
  describe('when commit has an associated pull request', () => {
    let requestSpy: jasmine.Spy;
    let res: CommitSelected[];
    beforeEach(async () => {
      requestSpy = spyOn(gqlRequest, 'gqlRequest').and.returnValues(
        { user: { id: 'myUserId' } },
        commitsWithPullRequestsMock
      );

      const options = getDefaultOptions();
      res = await fetchCommitsByAuthor(options);
    });

    it('Should return a list of commits with pullNumber and existing backports', () => {
      expect(res).toEqual([
        {
          sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
          message: 'Add ðŸ‘» (2e63475c)',
          existingBackports: []
        },
        {
          sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
          message: 'Add witch (#85)',
          pullNumber: 85,
          existingBackports: []
        },
        {
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
          message: 'Add SF mention (#80)',
          pullNumber: 80,
          existingBackports: [{ branch: '6.3', state: 'MERGED' }]
        },
        {
          sha: '3827bbbaf39914eda4f02f6940189844375fd097',
          message: 'Add backport config (3827bbba)',
          existingBackports: []
        },
        {
          sha: '5ea0da550ac191029459289d67f99ad7d310812b',
          message: 'Initial commit (5ea0da55)',
          existingBackports: []
        }
      ]);
    });

    it('should call with correct args to fetch author id', () => {
      expect(requestSpy.calls.argsFor(0)).toMatchSnapshot();
    });

    it('should call with correct args to fetch commits', () => {
      expect(requestSpy.calls.argsFor(1)).toMatchSnapshot();
    });
  });

  describe('when a custom github api hostname is supplied', () => {
    it('should be used in gql requests', async () => {
      const requestSpy = spyOn(gqlRequest, 'gqlRequest').and.returnValues(
        { user: { id: 'myUserId' } },
        commitsWithPullRequestsMock
      );

      const options = getDefaultOptions({
        apiHostname: 'api.github.my-company.com'
      });
      await fetchCommitsByAuthor(options);

      expect(requestSpy.calls.argsFor(0)[0].apiHostname).toBe(
        'api.github.my-company.com'
      );
      expect(requestSpy.calls.argsFor(1)[0].apiHostname).toBe(
        'api.github.my-company.com'
      );
    });
  });
});

describe('getExistingBackportPRs', () => {
  let pullRequest: PullRequestEdge;
  beforeEach(() => {
    pullRequest = {
      node: {
        number: 1234,
        timelineItems: {
          edges: [
            {
              node: {
                source: {
                  __typename: 'PullRequest',
                  state: 'MERGED' as const,
                  commits: {
                    edges: [
                      { node: { commit: { message: 'my message (#1234)' } } }
                    ]
                  },
                  baseRefName: '7.x'
                }
              }
            }
          ]
        }
      }
    };
  });

  it('should return a result when commit messages match', () => {
    const existingPRs = getExistingBackportPRs(
      'my message (#1234)',
      pullRequest
    );

    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });

  it('should return a result when first line of commit message matches', () => {
    pullRequest.node.timelineItems.edges[0].node.source.commits.edges[0].node.commit.message =
      'my message (#1234)\n\nsomething else';
    const existingPRs = getExistingBackportPRs(
      'my message (#1234)',
      pullRequest
    );

    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });
});
