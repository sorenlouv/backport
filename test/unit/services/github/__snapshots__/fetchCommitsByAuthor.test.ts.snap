// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch author id 1`] = `
Array [
  Object {
    "accessToken": "myAccessToken",
    "apiHostname": "api.github.com",
    "query": "
    query getIdByLogin($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
    "variables": Object {
      "login": "sqren",
    },
  },
]
`;

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch commits 1`] = `
Array [
  Object {
    "accessToken": "myAccessToken",
    "apiHostname": "api.github.com",
    "query": "
    query getCommitsByAuthorQuery(
      $repoOwner: String!
      $repoName: String!
      $commitsCount: Int!
      $authorId: ID
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: \\"master\\") {
          target {
            ... on Commit {
              history(first: $commitsCount, author: { id: $authorId }) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          number
                          timelineItems(
                            last: 20
                            itemTypes: CROSS_REFERENCED_EVENT
                          ) {
                            edges {
                              node {
                                ... on CrossReferencedEvent {
                                  source {
                                    __typename
                                    ... on PullRequest {
                                      state
                                      baseRefName
                                      commits(first: 20) {
                                        edges {
                                          node {
                                            commit {
                                              message
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  ",
    "variables": Object {
      "authorId": "myUserId",
      "commitsCount": 10,
      "repoName": "kibana",
      "repoOwner": "elastic",
    },
  },
]
`;
